services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/extension/fixture/rule/CleanupSwitchesRule.groovy

```diff
@@ -25,24 +27,66 @@ class CleanupSwitchesRule extends AbstractRuleExtension<CleanupSwitches> {
-                    assert rules.properRules.empty
+                    rules.excessRules.empty ?: failWith(new ExcessRulesException(
+                            "$sw.dpId switch appear to have excess rules at the end of the test.", rules))
+                    rules.properRules.empty ?: failWith(new ProperRulesException(
+                            "FATAL: there are proper rules left after the test. This usually means" +
+                                    " that certain flow was not properly removed from database and Kilda is in an" +
+                                    " inconsistent state. Aborting.\nswitch: $sw.dpId", rules))
+                    rules.missingRules.empty ?: failWith(new MissingRulesException(
+                            "FATAL: there are missing rules left after the test. This usually means" +
+                                    " that certain flow was not properly removed from database and Kilda is in an" +
+                                    " inconsistent state. Aborting.\nswitch: $sw.dpId", rules))
+                }
+            } catch (WaitTimeoutException e) {
+                //TODO(rtretiak): Not able to deal with randomly appearing rules from previous tests at this point.
+                //Waiting for new Kilda patch with Hub&Spoke features
+                def originalError = e.originalError
+                switch (originalError) {
+                    case ExcessRulesException:
+                        log.warn("Force switch cleanup.", originalError)
+                        northbound.deleteSwitchRules(sw.dpId, DeleteRulesAction.IGNORE_DEFAULTS)
+                        break
+                    case MissingRulesException:
+                    case ProperRulesException:
+                        //we are unable to deal with this. Kilda is inconsistent.
+                        throw originalError
+                    default:
+                        throw originalError
+
                 }
             }
-        } catch (WaitTimeoutException e) {
-            //TODO(rtretiak): Not able to deal with randomly appearing rules from previous tests at this point.
-            //Waiting for new Kilda patch with Hub&Spoke features
-            log.warn("Switches do not appear to be cleaned at the end of the test. Force switches cleanup." +
-                    "\n\nOriginal error:\n\n$e")
-            topology.activeSwitches.each { sw ->
-                northbound.deleteSwitchRules(sw.dpId, DeleteRulesAction.IGNORE_DEFAULTS)
-            }
         }
     }
+
+    static void failWith(Throwable t) {
+        throw t
+    }
+
+    private static abstract class RulesException extends RuntimeException {
+        RulesValidationResult rules
+
+        RulesException(String message, RulesValidationResult rules) {
+            super(message)
+            this.rules = rules
+        }
+
+        @Override
+        String toString() {
+            super.toString() + "rules: $rules"
```

Comment:
Please, capitalize 'r'.

Reply:
Also, I suggest deleting this toString() and add printing rules to super(), L75. You did the same in Wrappers.

super(message + "\nRules: $rules")